### ISSUE TITLE
TF (v1.9+) unable to call subprocess when making a tf.data.dataset during training in an estimator

### ISSUE THREAD

**Issue by SumNeuron** [tensorflow/tensorflow#29359]
What I am trying to do:

I have a TSV file consisting of string and integer values.

Each line in the file corresponds to a record e.g. 

```tsv
a  1  2    # record 1
b  3  4    # record 2
```

The values of these records are a condensed way of storing part of the input data. 

I have a bash function that can take the stringified tsv file, e.g.

```bash
my-bash-func "a\t\1\t2\nb\t3\t4\n"
```

which will return the input for my model.

I have a python wrapper function which uses `subprocess` to pipe the file to this bash command and return the output to python:

```python
process(['my-bash-func'], stdin="a\t\1\t2\nb\t3\t4\n")
```

where process might be defined as:

```python
def process(command:list, stdin:str, popen_options={}):
    '''
    Arguments:
        command (list): a list of strings indicating the command and its
            arguments to spawn as a subprocess.

        stdin (str): passed as stdin to the subprocess. Assumed to be utf-8
            encoded.

        popen_options (dict): used to configure the subprocess.Popen command

    Returns:
        stdout, stderr
    '''
    command = clean_command(command)
    popen_config = POPEN_DEFAULTS.copy()
    popen_config.update(popen_options)
    try:
        pid = subprocess.Popen(args = command, **popen_config)
        stdout_data, stderr_data = pid.communicate(stdin)
    except OSError as err:
        error_message(command, err)
        sys.exit(1)
    if pid.returncode != 0:
        error_message(command, 'pid code {}'.format(pid.returncode), stdout_data, stderr_data)
    return stdout_data, stderr_data
```

an example output might be:

```python
"x\t\y\t3\nw\tv\t8\n"
```

after parsing, splitting lines and tabs, the constructed input is now:

```python
a  1  2  x  y  3  # record 1
b  3  4  w  v  8 # record 2
```

with this information both the input to the model and the label for the input can be constructed, via a python function:

```
record = ["a", 1, 2, "x", "y", 3]
(features, label) = restored_record_to_feature_label_tuple(record)
```
for _each_ epoch during training some randomness needs to be injected. This is done via altering the values in the TSV e.g. record 1 (from the TSV) might have the following values over three epochs:

```python
a   1  2    # record 1 from TSV on epoch 1
a  -1  3    # record 1 from TSV on epoch 2
a   2  1    # record 1 from TSV on epoch 3
```

even with `tf.py_func` I have been unable to find a way to string together this input pipeline:

```python
# pseudo-code

# only needs to be done once. This TSV can also be encoded as tf Records.
tsv_data = parse_tsv(read_tsv(filename))

# done once per epoch
tsv_data_plus_randomness = randomify(tsv_data)

# reconstruct input (calls subprocess)
reconstructed_input = parse_tsv(
    process(
        ["my-bash-func"], 
        stringify(tsv_data_plus_randomness)
    )
)


features, labels = restored_record_to_feature_label_tuple(reconstructed_input)
```

Ideally the above code can become a function compatible with tf.Estimator api

```python
train_fn = # everything after reading in the tsv the first time

train_spec = tf.estimator.TrainSpec(input_fn=train_fn)
```

I _understand_ that this amount of processing is a bottleneck for the model. Ideally, if there were 100 epochs, the data for the 100 epochs of randomness would be pre-calculated and stored. Due to the nature of the model and the usefulness going forward, being able to work with the TSV (at the cost of training time) is worth the trade off. I state this after having done this approach already.

I have a reusable custom estimator [template][reusable estimator] which might be of use for trying to figure out why this is not possible. 

It seems that once the data is a tf.data.Dataset (be if from `tf.data.experimental.CsvDataset` or `tf.data.TFRecordDataset`), this amount of pythonic interface is not possible. 

-------------------------------

Addendum: 
I have been very open about what I know and do not know regarding Tensorflow. I have read extensively the documentation and even produced a concept to reduced the redundancy of the tf record interface via [FIO][https://pypi.org/project/fio/].

What I propose above is not an outrageous input pipeline and shouldn't be so obscure in regards to figuring out how to implement it. 

I went on quite an adventure trying to achieve a [reusable estimator]. While tf.estimator is "supported" (not actively developed but pre-made estimators may be added) in tf 2.0, it is likely to be removed or completely deprcyated by tf 3.0. Below includes a sampling of the adventure, and it is important to note, that even with some questions having >1k views, many are unanswered. Further, several issues (e.g. the issue regarding exporting device placement https://github.com/tensorflow/tensorflow/issues/23834) took so long to get any attention by developers that now that TF 2.0 is just about here, no one really cares to invest too much time into actually making user estimators functional as they will be replaced with keras models. 

Despite the colabs linked in these SO questions have been quite useful and individuals besides myself have used them (even apparently a company in china), sparing themselves the misery of figuring out how to get a basic custom model going (with some quality of life bells and whistles). 

I state all this to emphasize that I would not be asking this if I could figure out how to do it from the existing TF documentation / source code.

some of the relevant S.O.  questions:

- https://stackoverflow.com/questions/53634736/tensorflow-estimator-api-use-numpy-functions
- https://stackoverflow.com/questions/53414168/tensorflow-exportoutputs-predictouput-and-specifying-signature-constants-defau
- https://stackoverflow.com/questions/53410469/tensorflow-estimator-servinginputreceiver-features-vs-receiver-tensors-when-and
- https://stackoverflow.com/questions/53409652/tensorflow-estimator-clear-deivces-in-exporters
- https://stackoverflow.com/questions/53409547/tensorflow-estimator-bestexporter-event-file-pattern-doesnt-do-anything
- https://stackoverflow.com/questions/53356558/tensorflow-v1-10-load-savedmodel-with-different-device-placement-or-manually-se
- https://stackoverflow.com/questions/53356029/tensorflow-v1-10-serving-custom-estimator
- https://stackoverflow.com/questions/53355055/tensorflow-estimator-to-tensorflow-js
- https://stackoverflow.com/questions/53317235/tensorflow-custom-estimators-defining-estimator-spec-triggers-error
- https://stackoverflow.com/questions/53307954/tensorflow-custom-estimator-predict-throwing-value-error
- https://stackoverflow.com/questions/53226898/tensorflow-custom-estimator-stuck-when-calling-evaluate-after-training
- https://stackoverflow.com/questions/52874647/tensorflow-v1-10-why-is-an-input-serving-receiver-function-needed-when-checkpoi
- https://stackoverflow.com/questions/52641737/tensorflow-1-10-custom-estimator-early-stopping-with-train-and-evaluate
- https://stackoverflow.com/questions/52064866/tensorflow-1-10-tfrecorddataset-recovering-tfrecords
- https://stackoverflow.com/questions/52035692/tensorflow-v1-10-store-images-as-byte-strings-or-per-channel

[reusable estimator]: https://colab.research.google.com/drive/1GKAqEo7qSr6kMAgxPrOrudA5eLndQ6Ub#scrollTo=L8V8lpVFdyWL

**Comment by SumNeuron** [comment]
@gadagashwini any ideas?

**Comment by SumNeuron** [comment]
@ymodak I'd be curious to know your thoughts

**Comment by SumNeuron** [comment]
@tanzhenyu  thoughts?

**Comment by sachinprasadhs** [comment]
You seem to be using older version(1.x) of Tensorflow which is not supported anymore. Please try the latest [Tensorflow version](https://www.tensorflow.org/install/pip) and let us know if the problem still persists. Thanks!
